proposed/swift-interop.md

```diff
@@ -130,9 +130,15 @@ CoreCLR and NativeAOT currently block the `VectorX<T>` types from P/Invokes as t
 
 ##### Automatic Reference Counting and Lifetime Management
 
-Swift has a strongly-defined lifetime and ownership model. This model is specified in the Swift ABI and is similar to Objective-C's ARC (Automatic Reference Counting) system. When .NET calls into Swift, the .NET GC is responsible for managing all managed objects. Unmanaged objects from C# should either implement `IDisposable` or utilize a designated thin wrapper over the Swift memory allocator, currently accessible through the `NativeMemory` class, to explicitly release memory. It's important to ensure that when a Swift callee function allocates an "unsafe" or "raw" pointer types, such as UnsafeMutablePointer and UnsafeRawPointer, where explicit control over memory is needed, and the pointer is returned to .NET, the memory is not dereferenced after the call returns. Also, if a C# managed object is allocated in a callee function and returned to Swift, the .NET GC will eventually collect it, but Swift will keep track using ARC, which represents an invalid case and should be handled by projection tools.
+Swift has a strongly-defined lifetime and ownership model. This model is specified in the Swift ABI and is similar to Objective-C's ARC (Automatic Reference Counting) system. When .NET calls into Swift, the .NET GC is responsible for managing all managed objects.
 
-The Binding Tools for Swift tooling handles these explicit lifetime semantics with generated Swift code. In the new Swift/.NET interop, management of these lifetime semantics will be done by the Swift projection layer and not by the raw calling-convention support. If any GC interaction is required to handle the lifetime semantics correctly, we should take an approach more similar to the `ComWrappers` support (higher-level, less complex interop interface) rather than the Objective-C interop support (lower-level, basically only usable by the ObjCRuntime implementation).
+There are two strategies for managing native memory in .NET: destructors and `IDisposable`. Destructors abstract away memory management from the user and are managed by the GC. They provide a way to release unmanaged resources when an object is collected by the GC. `IDisposable` offers an explicit mechanism for releasing unmanaged resources with deterministic control over when resources are released. The preferred behavior for general cases would be to implement destructors. This approach aligns with the .NET pattern and offers codegen benefits by avoiding excessive `using` statements. Ideally, the tooling should only use `IDisposable` for custom deinit.
```

Comment:
What are the compatibility rules for adding/removing custom deinit in Swift?

Are introduction or removal of a custom deinit considered to be a compatible change in Swift? If they are compatible changes, it would be problematic to map custom deinit to IDIsposable. Adding/removing IDisposable is a breaking change in .NET.

Reply:
This is unfortunate -- I bet Swift is fine with removing a deinit since there's always an implicit deinit that it can fall back on.

I think the fundamental problem here is that Swift does not have a distinction between "deterministic" destruction and non-deterministic destruction. In contrast, C# does. And, importantly, that distinction is very important in C#. If you require deterministic destruction then timely destruction may be a requirement for your application, where delaying destruction may cause your application to fail. Conversely, if you don't require it, using deterministic destruction (`IDisposable`/using) for all resources is very expensive and will likely cause significant performance degradation. It's also a large programmer burden as disposability is transitive and any IDisposable fields require the parent type to also be IDisposable.

"Custom deinit" is a heuristic that attempts to map between C#, which differentiates the two concepts, and Swift, which doesn't. But since there is no way to determine statically whether a Swift type actually requires deterministic destruction, I don't think there's any heuristic that will be perfectly accurate.

That said, maybe we can do better than "custom deinit". Suggestions welcome.

Reply:
Does this imply that we should have a solid reason for implementing the `IDisposable` pattern on a type as introducing or removing explicit memory management via `IDisposable` could lead to breaking changes?

This is the case where direct mapping is not possible, and introducing `IDisposable` only when absolutely necessary can help minimize breaking changes by switching to the explicit memory management.

Reply:
> This is the case where direct mapping is not possible, and introducing IDisposable only when absolutely necessary

Another option here would be to follow the built-in COM mechanism and add a helper function like [`Marshal.FinalReleaseComObject()`](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.marshal.finalreleasecomobject). This could be on `Marshal` or another more targeted type, and use any number of tricks to provide efficient deterministic release semantics. This has the added benefit of not forcing users to litter their code with type checks/casts.

Note that `ComWrappers` did use `IDisposable` for specific COM scenarios. In that case though the user is driving the experience by explicitly requesting the deterministic support - see [this section](https://learn.microsoft.com/dotnet/standard/native-interop/tutorial-comwrappers#step-4--handle-native-object-wrapper-lifetime-details) of the `ComWrappers` tutorial.

Reply:
I'll offer a datapoint here that there are at least 4 issues at play here:
- Swift has very different semantics than C# for value types. Value types can induce code to get executed when they go out of scope (for example, a value type contains a reference type).
- Swift maintains two different reference counts: a strong reference count and a weak reference count on heap-allocated types, not a simple reference count. The weak reference count is always used implicitly at allocation time (IIRC - it's been a while since I put this under a magnifying glass) and may be used explicitly by assigning to a member declared as `weak`
- At binding time, it is not always deterministic that a value type will always be blittable (for example, a non-frozen struct could get a member added in the future) and this will create a compatibility issue for us in the future

In BTfS, when we are presented with a Swift object, the C# binding implements `IDisposable` and takes a swift weak reference to the object via the swift runtime routine `swift_unownedRetain` and keep a weak GCHandle. `Dispose (false)` gets called by the finalizer.

Initially, I tried to draw a distinction between blittable and non-blittable value types, but things that make that challenging were all the special cases: value types that contain mutating methods or properties, `inout` parameters, frozen/not frozen, opaque layout of enums etc. It was an endless supply of bugs in the marshaling. My solution for this was to make all values type handling uniform.

Not being an expert on it, does COM interop exist on non-windows platforms?

I'm pinging @rolfbjarne here as well since he was responsible for the handling of ObjC mapping, which required similar work.

Reply:
> In BTfS, when we are presented with a Swift object, the C# binding implements IDisposable and takes a swift weak reference to the object via the swift runtime routine swift_unownedRetain and keep a weak GCHandle. Dispose (false) gets called by the finalizer.

This describes the implementation, but what was the API? Were users expected to always call `IDisposable`? If so, how did this affect performance? If not, are there cases where delaying the cleanup to the GC could cause resource exhaustion? How did you deal with analyzers that warn if all IDisposable instances are not disposed?

Reply:
In our ObjC interop, all managed wrappers for ObjC objects are IDisposable. The decision was made before my time, but here are a few points:

* It makes it possible for developers to make the behavior deterministic. They need to be able to free native resources when needed, we can't really say "just wait for the GC to kick in", because by the time the GC kicks in, the app might have been terminated by the OS for using too much memory (note that mobile devices don't have virtual memory like desktops do).
* We could try to only make potentially big objects IDisposable, but IMHO that's a rabbit hole. There's no limit to how big a simple string can be for instance. We also don't know the internals of how Swift types are implemented (and that can also change in the future).
* While there have been rare cases of projects running into performance problems with our object lifetime implementation (Visual Studio for Mac itself was probably the biggest), there have been *many* more projects with memory problems that were fixed by just disposing objects in a timely manner.

> Without knowing details of the type, it seems impossible for me, the caller, to understand whether I should or shouldn't dispose it.

A counter-point here is that if we don't make a type disposable, we're making the decision for the caller, while being potentially even less knowledgeable about the type.

One goal for the swift interop tooling is for the tooling to be as automatic/autuomated as possible, if we have to read through documentation in order to properly bind an API, then it makes the binding process much more slower, in particular because there might not be much documentation for new APIs until way after they've been released. At that point the caller will know much more about it than we did when we bound it.

IMHO making everything disposable isn't necessarily a good solution, but I think the alternative is worse.

Reply:
> IMHO making everything disposable isn't necessarily a good solution, but I think the alternative is worse.

The alternative in this case being a low-level static function that can be used to dispose of instances at the consumers discretion?

Reply:
> Not being an expert on it, does COM interop exist on non-windows platforms?

@stephen-hawley It does through the `ComWrappers` API. It doesn't rely upon the `Marshal.FinalReleaseComObject()` API as that is only for built-in COM, Windows only. The `ComWrappers` API recommends using `IDisposable` for narrow scenarios where determinism is a hard requirement. For all other scenarios the GC should be relied upon to clean up when appropriate.

I think, could be misremembering, that the non-frozen scenarios aren't planned, is that right @jkoritzinsky?

Perhaps a way to get to an answer would be to see some Objective-C user applications and note how common the leveraging of the `IDisposable` semantics was. If it was relied upon a majority of the time, where a different application wasn't appropriate, then sure it is common enough that we should place it on the types. However, if we observe there are relatively few instances where it would actually be needed, then a static function would seem more appropriate.

Reply:
Non-frozen scenarios are planned (most of CryptoKit itself is non-frozen). Frozen is the only case that needs ABI work in the runtime, all the work is in the projections case for everything else.

Reply:
> > In BTfS, when we are presented with a Swift object, the C# binding implements IDisposable and takes a swift weak reference to the object via the swift runtime routine swift_unownedRetain and keep a weak GCHandle. Dispose (false) gets called by the finalizer.
> 
> This describes the implementation, but what was the API? Were users expected to always call `IDisposable`? If so, how did this affect performance? If not, are there cases where delaying the cleanup to the GC could cause resource exhaustion? How did you deal with analyzers that warn if all IDisposable instances are not disposed?

The API is effectively `IDisposable`. Every heap allocated type extends from a common base class which handles the details for you.

In my experience, the more we do for the users the better off we'd be. We can say all we want, "you need to be aware of object lifetimes", but in reality few customers will actually do that. That said, the approach I used in BTfS was to use and interface, `ISwiftObject` to define how the object would behave (and in this case, all it has is an `IntPtr` for the object handle) but use a common base class for typical implementations (`SwiftNativeObject`). No code that figures out what to do with an object uses `SwiftNativeObject` but instead uses `ISwiftObject`. For example, in run-time marshaling an arbitrary C# object to swift, I check to see if it implements `ISwiftObject` as one step. Because of that distinction between the interface and base class, making our typical bindings `IDisposable` doesn't preclude a user from defining something that isn't.

Swift trivia: swift, unlike C#, Java, etc, doesn't have an 'object` base class. Generally, swift takes a lot of steps to keep you from doing runtime type checks. Implicitly, every heap allocated object implements the `AnyObject` protocol.

Reply:
> > IMHO making everything disposable isn't necessarily a good solution, but I think the alternative is worse.
> 
> The alternative in this case being a low-level static function that can be used to dispose of instances at the consumers discretion?

Ah, that might work, I was thinking about not providing a way to dispose of instances at all.

Reply:
> Perhaps a way to get to an answer would be to see some Objective-C user applications and note how common the leveraging of the `IDisposable` semantics was.

It's not all that common, because there's a major downside to disposing in our implementation: it partially disconnects the managed object and the native object from eachother, which may become a problem if the native object shows up in managed code again for whatever reason. In that case we'll check if there's a corresponding managed object for the native object, won't find anything (because they got disconnected), then try to create a new managed object, which may fail (if there's no appropriate constructor).

On the other hand, when dealing with media (images, video and audio), disposing the data is pretty much required, because memory usage goes up really fast. A photo is surprisingly big on iPhones; it doesn't take many of them to reach the memory limit on iOS.

Disposing can also be used sometimes to break reference cycles, although typically other solutions are better (nulling out properties/fields for instance).

Our recommendation is to only dispose object if you know you have to, otherwise leave the GC to do its work.

Reply:
> Our recommendation is to only dispose object if you know you have to, otherwise leave the GC to do its work.

In this context, if a Swift type is mapped to a struct or class on the .NET side, the corresponding interface should implement `IDisposable`. This allows users to explicitly dispose of memory if needed. In all other cases, implicit disposal will be handled by the GC and finalizers. 

Does such a pattern prevent codegen bloat and excessive usage of explicit disposal?

Reply:
Let's review some examples of existing bindings to gain a deeper understanding of how the `IDisposable` pattern is utilized. I've done a review of the [Xamarin Google APIs for iOS Components](https://github.com/xamarin/GoogleApisForiOSComponents) and [Google Mobile Unity Ads](https://github.com/googleads/googleads-mobile-unity), focusing specifically on the following applications: cloud messaging, mobile ads, analytics, and crashlytics. The goal of this review was to identify the use cases of IDisposable patterns.

In the existing Xamarin ObjC interop, the `NSObject` represents a base object and implements IDisposable.

### Cloud Messaging

In the [cloud messaging bindings](https://github.com/xamarin/GoogleApisForiOSComponents/blob/main/source/Firebase/CloudMessaging/ApiDefinition.cs), the following APIs require implementing the IDisposable pattern:
- `MessageInfo`
- `MessagingDelegate`
- `Messaging`
- `MessagingExtensionHelper`

In the [sample](https://github.com/xamarin/GoogleApisForiOSComponents/blob/main/samples/Firebase/CloudMessaging/CloudMessagingSampleLegacy/AppDelegate.cs) application, these types are used in AppDelegate, but there are no explict disposing of objects.

There is another example where IDisposable pattern is used to remove notification handles

```csharp
var sendSuccessToken = Messaging.Notifications.ObserveSendSuccess (HandleSendMessageSuccess);
var sendErrorToken = Messaging.Notifications.ObserveSendError (HandleSendMessageError);

// Call this lines to stop receiving notifications
sendSuccessToken.Dispose ();
sendErrorToken.Dispose ();

// Another option: Calling native code could explicitly release observers.

var sendSuccessToken = NSNotificationCenter.DefaultCenter.AddObserver (Messaging.SendSuccessNotification, HandleSendMessageSuccess);
var sendErrorToken = NSNotificationCenter.DefaultCenter.AddObserver (Messaging.SendSuccessNotification, HandleSendMessageError);

// Call this lines to stop receiving notifications
NSNotificationCenter.DefaultCenter.RemoveObserver (sendSuccessToken);
NSNotificationCenter.DefaultCenter.RemoveObserver (sendErrorToken);
```

Analytics

In the [analytics bindings](https://github.com/xamarin/GoogleApisForiOSComponents/tree/main/samples/Firebase/Analytics/AnalyticsSample), the following APIs require implementing the IDisposable pattern:
- `Analytics`
- `DictionaryBuilder`
- `Fields`
- `Logger`
- `TrackedViewController`
- `Tracker`
- `EcommerceFields`
- `EcommerceProduct`
- `EcommerceProductAction`
- `EcommercePromotion`

In the [samples](https://github.com/xamarin/GoogleApisForiOSComponents/tree/main/samples/Firebase/Analytics/AnalyticsSample), they are not explicitly released.

MobileAds

In the [mobile ads bindings](https://github.com/xamarin/GoogleApisForiOSComponents/tree/main/source/Google/MobileAds), there are 43 interfaces with base types implementing IDisposable section.

There are 6 instances of disposal in the [sample application](https://github.com/xamarin/GoogleApisForiOSComponents/tree/main/samples/Google/MobileAds):
```csharp
// You need to explicitly Dispose Interstitial when you dont need it anymore
// to avoid crashes if pending request are in progress
void RemoveAdFromTableView ()
{
	if (adViewTableView != null) {
		if (adOnTable) {
			dvcDialog.Root.RemoveAt (idx: 2, anim: UITableViewRowAnimation.Fade);
		}
		adOnTable = false;

		// You need to explicitly Dispose BannerView when you dont need it anymore
		// to avoid crashes if pending request are in progress
		adViewTableView.Dispose();
		adViewTableView = null;
	}
}
```

### GoogleAds Mobile Unity

In the [google mobile bindings](https://github.com/googleads/googleads-mobile-unity/tree/main/source/plugin/Assets/GoogleMobileAds), the following types implement IDisposable, but none of them are disposed explicitly.
 - `AdManagerBannerClient`
 - `AppOpenAdClient`
 - `BannerClient`
 - `InterstitialClient`
 - `NativeOverlayAdClient`
 - `RewardedAdClient`
 - `RewardedInterstitialAdClient`

### Crashlytics

In the [crashlytics bindings](https://github.com/xamarin/GoogleApisForiOSComponents/blob/main/source/Firebase/Crashlytics/ApiDefinition.cs), the following APIs require implementing IDisposable:
- `Crashlytics`
- `ExceptionModel`
- `StackFrame`

In the [sample application](https://github.com/xamarin/GoogleApisForiOSComponents/tree/main/samples/Firebase/Crashlytics), none of them are disposed explicitly.


### Xamarin Designer

The Xamarin Designer (deprecated) for iOS is a visual designer that contain controls with IDisposable pattern. Samples from the mentioned bindings that implement Xamarin Designer templates contain a ReleaseDesignerOutlets function that explicitly releases all unmanaged references. Here is an example:
```csharp
[Outlet]
UIKit.UILabel LblPreview { get; set; }

[Outlet]
UIKit.UILabel LblSubtitle { get; set; }

[Outlet]
UIKit.UILabel LblTitle { get; set; }

void ReleaseDesignerOutlets ()
{
	if (LblPreview != null) {
		LblPreview.Dispose ();
		LblPreview = null;
	}

	if (LblSubtitle != null) {
		LblSubtitle.Dispose ();
		LblSubtitle = null;
	}

	if (LblTitle != null) {
		LblTitle.Dispose ();
		LblTitle = null;
	}
}
```

The collected data could help us move this discussion/decision forward.

Reply:
Thanks @kotlarmilos, this is a lot of great work!

Based on real-world examples, it doesn't seem like people actually use Dispose in most cases. Based on that, I would recommend using a static method like Marshal.ReleaseComObject (or even a different interface if necessary) rather than adding a bunch of ceremony to the API, but that's just my opinion.

My main concern is accurately providing guidance to authors on whether or not manual disposal is expected, and it seems like it's not.

@jkoritzinsky @AaronRobinsonMSFT @jkotas thoughts?

Reply:
Something like `Dispose` is fine but I might be careful about using `Dispose` itself because it could be confused with IDisposable and raise the same problems. Not implementing the interface is a kind of subtle shift that doesn't seem to present a lot of clarity.

> Implementing IDisposable on the projected types can trigger analyzer warnings- for example [CA2000](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2000), or [CA2213](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca2213) - although these analyzers are not enabled by default currently (that said it seems that other tools do report similar diagnostics - Resharper and so on)

This is indeed my main concern -- some `IDisposable` types require disposal for correct usage. These don't. It's really hard to tell the ones that do from the ones that don't, so my general recommendation for people is to avoid `IDisposable` if it's not intended to be used in normal operation.